Correct DateColumn Format in Edit Mode

tortaepaisitphongse opened the pull request:
## Describe your changes

This PR fixes Issue #12850 where `st.column_config.DateColumn` always defaulted to MM/DD/YYYY format during text editing, regardless of the `format` parameter specified. The format parameter only affected display mode, not edit mode, causing confusion for users who saw dates in one format but had to edit them in another.

**Implementation:**

- Created a custom date editor component (`DateTextCellEditor`) that provides format-aware text input for date cells
- Modified `DateTimeColumn` to detect custom date formats and store them in a `userFormat` field within cell data
- Integrated the custom editor into `useCustomEditors` hook to override the default editor for date cells with custom formats
- The custom editor uses moment.js for strict date parsing according to the user-specified format and provides real-time validation feedback

**Technical Approach:**
The solution implements a custom editor wrapper that intercepts editing for date cells with custom formats, leveraging `glide-data-grid`'s `provideEditor` callback mechanism. This approach was chosen because the underlying `glide-data-grid-cells` library only supports predefined input types ("date", "time", "datetime-local") and doesn't support custom format masks.

**Files Changed:**

- `frontend/lib/src/components/widgets/DataFrame/columns/cells/DateTextCellEditor.tsx` (new)
- `frontend/lib/src/components/widgets/DataFrame/columns/DateTimeColumn.ts` (modified)
- `frontend/lib/src/components/widgets/DataFrame/hooks/useCustomEditors.ts` (modified)
- `frontend/lib/src/components/widgets/DataFrame/columns/DateTimeColumn.test.ts` (modified - added tests)
- `frontend/lib/src/components/widgets/DataFrame/hooks/useCustomEditors.test.ts` (modified - added tests)

## Screenshot or video (only for visual changes)

## GitHub Issue Link (if applicable)

Fixes #12850

## Testing Plan

### Unit Tests (JS and/or Python)

**JavaScript Unit Tests:**

- `DateTimeColumn.test.ts`: Added 4 new test cases for `userFormat` field behavior

  - Tests verify `userFormat` is set for custom formats (DD.MM.YYYY, MM/DD/YYYY, YYYY-MM-DD)
  - Tests verify `userFormat` is set to default format (YYYY-MM-DD) when no format is provided
  - Tests verify `userFormat` is not set for special formats (localized, distance, iso8601)
  - Tests verify `userFormat` is set even when explicitly provided format matches default
  - All 92 tests passing (including existing tests)

- `useCustomEditors.test.ts`: Added 4 new test cases for custom editor selection
  - Tests verify `DateTextCellEditor` is returned for date cells with `userFormat`
  - Tests verify `undefined` is returned for date cells without `userFormat`
  - Parameterized test for multiple date formats
  - All 9 tests passing (including existing tests)

**Test Results:**

```
DateTimeColumn.test.ts: 92/92 tests passing
useCustomEditors.test.ts: 9/9 tests passing
```

### E2E Tests

No E2E tests added. The existing E2E test suite should continue to pass as this change only affects the edit mode behavior for date columns with custom formats, and the underlying data handling remains unchanged.

### Any manual testing needed?

Yes, manual testing was performed to verify the end-to-end user experience:

**Manual Testing Completed:**

- Verified edit mode matches display mode for DD.MM.YYYY format
- Verified edit mode matches display mode for MM/DD/YYYY format
- Verified edit mode matches display mode for YYYY-MM-DD format (default)
- Verified default format (no format parameter) works correctly
- Verified special formats (localized, distance, iso8601) still use default browser date picker
- Confirmed that saving works correctly after editing dates in custom formats
- Tested keyboard shortcuts (Enter to commit, Escape to cancel)

<img width="1440" height="852" alt="Screenshot 2025-12-06 at 6 05 35 PM" src="https://github.com/user-attachments/assets/2b6b0ef4-6ed8-4f90-b996-1f8ae9aad35b" />
<img width="1336" height="314" alt="Screenshot 2025-12-06 at 6 05 42 PM" src="https://github.com/user-attachments/assets/dd7c49a7-8de3-46ad-bb26-36a6d023612a" />
<img width="1334" height="288" alt="Screenshot 2025-12-06 at 6 05 48 PM" src="https://github.com/user-attachments/assets/dc653b7a-4988-46bf-bddd-0855b97f531c" />
<img width="1336" height="285" alt="Screenshot 2025-12-06 at 6 05 54 PM" src="https://github.com/user-attachments/assets/33cd1bab-3ce4-4c77-82b7-8265a8984db9" />
<img width="1359" height="293" alt="Screenshot 2025-12-06 at 6 06 00 PM" src="https://github.com/user-attachments/assets/dbca6ded-01e9-42e0-baf4-379d57dbb50f" />
<img width="1359" height="297" alt="Screenshot 2025-12-06 at 6 06 17 PM" src="https://github.com/user-attachments/assets/dbe017e7-0c6b-4bfe-97f2-e3a0560d8faa" />

**Static Analysis:**

- All code passes ESLint linting (no errors or warnings)
- All TypeScript type checks pass (no type errors)
- Code formatted with Prettier according to project standards

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

tortaepaisitphongse committed:
enhancement: respect DateColumn format parameter in edit mode (Issue #12850) with custom DateTextCellEditor component

tortaepaisitphongse committed:
test: add tests for DateColumn custom format editor

tortaepaisitphongse committed:
Merge branch 'develop' into feature/DateColumn_format

tortaepaisitphongse committed:
fix: linter error in useCustomEditors.test

tortaepaisitphongse committed:
Merge branch 'feature/DateColumn_format' of https://github.com/zachikeh123/streamlit into feature/DateColumn_format

tortaepaisitphongse committed:
fix: required cell implementation + unit tests + documentation edit + linter issues

tortaepaisitphongse committed:
Merge branch 'develop' into feature/DateColumn_format

tortaepaisitphongse commented:
## Code Review Fixes Addressed
- Add accessibility attributes (`aria-invalid`, `aria-describedby`, `role="alert"`) for error states
- Fix CSS `borderRadius` value (use `0` instead of `"none"`)
- Add `required` field support with proper validation logic
- Fix null value saving for non-required cells (pass `updatedCell` instead of `undefined`)
- Simplify blur behavior: restore original value and close editor on validation error
- Improve type safety: use type guards instead of `as any` assertions
- Update documentation comments for accuracy

## Testing
- Add comprehensive unit tests for `DateTextCellEditor` (26 tests covering parsing, validation, blur, keyboard shortcuts, accessibility, edge cases)
- Update `DateTimeColumn` tests to verify `userFormat` and `required` fields
- Update `useCustomEditors` tests for custom editor integration

sfc-gh-lwilby commented:
Hi @tortaepaisitphongse, 

Thank you for submitting this PR. Before doing a deeper review a couple of things stood out to me. 

1) Can you provide a description of the gap and an explanation of why you think your proposed solution is the best approach. Have you considered any alternatives? 

2) It looks like there is a lot of casting to `any`. If this is necessary, please justify why for each one because this goes against our best practices. 

3) It looks like in some places you aren't following our standard practices like using `styled-elements`. Please familiarize yourself with how we typically write components and follow a similar style. 

Thank you!

lukasmasuch commented:
@cursor review

tortaepaisitphongse commented:
Hi @sfc-gh-lwilby! Thank you for the thoughtful questions. Please let me know if you have any further questions or recommendations :)

## Gap Description and Solution Justification

### The Problem
The `st.column_config.DateColumn` in `st.data_editor` does not respect the `format` parameter in edit mode. While dates display correctly in the specified format (e.g., "DD.MM.YYYY", "MM/DD/YYYY"), when users click to edit a cell, the browser's native date picker appears with a browser-dependent format (typically MM/DD/YYYY), creating a confusing user experience where the display format differs from the edit format.

### Root Cause
The underlying `@glideapps/glide-data-grid-cells` library's `DatePickerCell` component only supports native HTML date input types (`"date"`, `"time"`, `"datetime-local"`), which do not accept custom format masks. The library's `DateKind` type is limited to these three values, and attempting to pass custom formats results in errors.

### Solution Approach
We implemented a custom text input editor (`DateTextCellEditor`) that:
- Replaces the native date picker for cells with custom formats
- Uses `moment.js` (already a dependency) for strict format parsing and validation
- Provides real-time validation feedback
- Maintains consistency between display and edit formats
- Uses existing infrastructure: Leverages `provideEditor` callback already used for JSON cells

### Alternatives Considered

1. **Modify the third-party library** (`@glideapps/glide-data-grid-cells`)
   - Would require forking and maintaining a fork
   - Risk breaking changes in upstream updates

2. **Use a different date picker library**
   - Would require significant refactoring of the entire date column implementation
   - Risk breaking existing functionality and increasing bundle size

## Justification for `as any` Type Assertions

## DateTimeColumn.ts (2 instances)

**Location 1: Line 176 (cellTemplate.data)**

```typescript
data: {
  // ... standard DatePickerCell props
  userFormat: formatToUse,
  required: props.isRequired,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
} as any,
```

**Justification**: The `DatePickerCellProps` type from `@glideapps/glide-data-grid-cells` does not include `userFormat` or `required` fields. These are custom fields we add to pass format information from the backend to the frontend custom editor. We cannot extend the third-party type definition without modifying their package. The `as any` assertion is necessary because to enable this functionality. 

**Location 2: Line 310 (getCell method)**

```typescript
data: {
  ...cellTemplate.data,
  date: cellData,
  displayDate,
  timezoneOffset,
  userFormat: formatToUse,
  required: props.isRequired,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
} as any,
```

**Justification**: Same as above.

### DateTimeColumn.test.ts (multiple instances)

**Justification**: Test files use `as any` to access the custom `userFormat` and `required` fields we added. The tests verify that our custom fields are correctly set, which requires accessing properties not in the original type.

### DateTextCellEditor.test.tsx (1 instance)

**Location: Line 63 (createMockCell helper)**

```typescript
data: {
  // ... standard props
  userFormat,
  required,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
} as any,
```

**Justification**: Test helper function needs to create mock cells with our custom fields. This is necessary for testing our custom editor functionality.

### DateTextCellEditor.tsx (1 instance - `as unknown`)

**Location: Line 96**

```typescript
const cellDataUnknown = cellData as unknown;
```

**Justification**: We use `as unknown` (not `as any`) as an intermediate step before applying a type guard (`hasUserFormat`). This is a safer pattern that allows us to check the structure at runtime and uses a type guard to narrow to our custom type.

## Styled Components Usage

Refactored `DateTextCellEditor` to use `@emotion/styled` components, following the same pattern as `ImageCellEditor.tsx` and `JsonViewer.tsx`.

**Changes made:**

- Replaced inline styles with styled components (`StyledEditorContainer`, `StyledInput`, `StyledError`)
- Used `$` prefix for transient props (following Emotion best practices) to pass theme values from glide-data-grid
- Styled components automatically access the emotion theme from context
- Merged provided theme (from glide-data-grid) with emotion theme via props

tortaepaisitphongse committed:
fix: UTC based date parsing + refactored theme usage/styled components

tortaepaisitphongse committed:
Merge branch 'develop' into feature/DateColumn_format

sfc-gh-lwilby commented:
OK, @tortaepaisitphongse , I think @lukasmasuch should weigh in on whether we want to go forward with this or instead modify the Glide data grid code. Lukas is also a maintainer of that library so it is a viable option to make changes there. To me, given all the typing overrides and amount of code added here it looks like modifying Glide might be a cleaner solution.

tortaepaisitphongse commented:
@sfc-gh-lwilby Got it, that definitely makes sense if editing the Glide data grid code is an option. Please let me know what the team decides to move forward with and thank you for your support!

jrieke commented:
Haven't tried this out yet but generally makes sense to me from product side and looks good on screenshots. Agree that we should move things to glide-data-grid as much as possible. Let me know how y'all decide to move forward with the implementation, then I can give this another look.

sfc-gh-lwilby commented:
@lukasmasuch can you weigh in on this one?

lukasmasuch commented:
I don't think its worth adding this complexity on Streamlit side just to enable a different format. The current input is based on the base HTML input that always adapts to the users local, which should be mostly fine for editing purpose. In case we go with a custom implementation, it would make sense to use the same date / datetime picker component that we use for `st.date_input`, `st.datetime_input`... However, this is a bigger project and might be better to wait until we made a decision if we stay in baseweb UI or not. And if we decide to support the format with that, the proper solution would be to make it work directly within glide-data-grid. Overall, I don't think we should fix this on Streamlit-side at the moment.

sfc-gh-lwilby commented:
@tortaepaisitphongse thank you for taking a look at this, but as Lukas explained above, we would rather go in a different direction and keep the changes in the glide-data-grid.
